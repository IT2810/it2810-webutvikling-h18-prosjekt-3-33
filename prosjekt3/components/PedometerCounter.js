import React from 'react';
import Expo from 'expo';
import { Pedometer } from 'expo';
import { StyleSheet, Text, View, TouchableOpacity } from 'react-native';
import * as Storage from './Storage';
import Colors from '../constants/Colors';

//Pedometer component used by StepGoal to get step data from core motion (iOS) or
//Google Fit (Android)
export default class PedometerCounter extends React.Component {
  state = {
    isPedometerActive: "checking",
    pastStepCount: 0,
    currentStepCount: 0,
    stepGoal: 0,
  };
  //Subscribe to pedometer count and setting state from props (from StepGoal)
  componentDidMount() {
    this._subscribe();
    this.setState({
      stepGoal: this.props.stepGoal
    })
  }
  //Unsubscribe from pedometer count on unmount
  componentWillUnmount() {
    this._unsubscribe();
  }
  //State changes on new props from StepGoal
  componentDidUpdate(prevProps, prevState){
    if(this.props.stepGoal != prevProps.stepGoal){
      this.setState({stepGoal: this.props.stepGoal})
    }
  }
//Subscribing to Core Motion and Google FIt
  _subscribe = () => {
    this._subscription = Pedometer.watchStepCount(result => {
      this.setState({
        currentStepCount: result.steps
      });
    });

    Pedometer.isAvailableAsync().then(
      result => {
        this.setState({
          isPedometerActive: String(result)
        });
      },
      error => {
        this.setState({
          isPedometerActive: "Could not get isPedometerActive: " + error
        });
      }
    );

    const end = new Date();
    const start = new Date();
    start.setDate(end.getDate() - 1);
    Pedometer.getStepCountAsync(start, end).then(
      result => {
        this.setState({ pastStepCount: result.steps });
      },
      error => {
        this.setState({
          pastStepCount: "Could not get stepCount: " + error
        });
      }
    );
  };

  _unsubscribe = () => {
    this._subscription && this._subscription.remove();
    this._subscription = null;
  };

  render() {
    //Renders a progress bar based on progress / goal relationship
    var totalProgress = this.state.pastStepCount + this.state.currentStepCount;
    var goal = this.state.stepGoal;
    return (
      <View style={styles.container}>
        <View style={styles.textContainer}>
          <Text>
            {totalProgress} out of {goal}
          </Text>
        </View>
        <View style={styles.progressBar}>
          <View style={progressBarStyle(totalProgress, goal)}>
          </View>
        </View>

      </View>
    );
  }
}
//Function for calculating the progressbar colors and width
progressBarStyle = function(progress, goal){
  let progressWidth = 0;
  let backgroundColor = '#000';
  let progressGoalRelation = progress/goal;
  if(progressGoalRelation>=1){
    progressWidth = 1;
    backgroundColor = '#75CDA8';
  }
  if (progressGoalRelation < 1 && progressGoalRelation > 0.4){
    progressWidth = progressGoalRelation;
    backgroundColor = '#EDBF2C';
  }
  if(progressGoalRelation<=0.4){
    progressWidth = progressGoalRelation;
    backgroundColor = '#DB0200';
  }
  return{
    flex: progressWidth,
    backgroundColor: backgroundColor,
  }
}
const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    marginTop: 15,
    alignItems: "center",
    justifyContent: "center"
  },
  progressBar: {
    flex: 1,
    flexDirection: 'row',
    width: 300,
    height: 20,
    backgroundColor: '#eee',
  },
  textContainer: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    width: 300,
  }
});

Expo.registerRootComponent(PedometerCounter);
