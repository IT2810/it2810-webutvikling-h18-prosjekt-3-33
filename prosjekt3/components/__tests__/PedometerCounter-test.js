import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import MobileButton from 'react-native';
import PedometerCounter from '../PedometerCounter';
import { Pedometer } from 'expo';


describe('PedometerCounter', () => {

	test('renders correctly', () => {
	  	const pedometer = renderer.create(<PedometerCounter />).toJSON();
	  	expect(pedometer).toMatchSnapshot();
	});

	test('handles subscription correctly', done => {
		expect.assertions(3);
		const pedometer = renderer.create(<PedometerCounter />).getInstance();
		expect(pedometer).toBeDefined();
		expect(pedometer.state.isPedometerActive).toEqual('checking');


		// doesn't work
		function callback() {
			// if error:
			    expect(pedometer.state.pastStepCount).not.toBe(0);
			    done();
		    // if success: 
		    	expect(pedometer.state.pastStepCount).toBeGreaterThanOrEqual(0)
		 }


		pedometer._subscribe(callback);
		console.log(pedometer.state);

	});
});