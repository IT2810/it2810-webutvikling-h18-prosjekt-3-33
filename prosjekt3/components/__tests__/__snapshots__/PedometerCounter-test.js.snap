// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PedometerCounter renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "marginTop": 15,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "width": 300,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      0
       out of 
    </Text>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#eee",
        "flex": 1,
        "flexDirection": "row",
        "height": 20,
        "width": 300,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#000",
          "flex": 0,
        }
      }
    />
  </View>
</View>
`;
